### PR TITLE
Spec dynamic_gather

### PR DESCRIPTION
**Pull request by sdasgup3**
fixes https://github.com/openxla/stablehlo/issues/2290

### CODE REVIEW

File: stablehlo/dialect/TypeInference.cpp [hunk 1]
```diff
@@ -1446,13 +1446,13 @@ static LogicalResult verifyGather(
   if (sliceSizesShape.hasStaticShape()) {
```
Comment (mlevesquedion) [inline stablehlo/dialect/TypeInference.cpp:1446]: I believe this check can be removed.

**Review by mlevesquedion — APPROVED** [review]
Looks good, just a few nits.

File: stablehlo/tests/interpret/dynamic_gather.mlir [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// RUN: stablehlo-translate --interpret -split-input-file %s
+
+func.func @gather_op_test() {
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply stablehlo/tests/interpret/dynamic_gather.mlir]: done

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -587,6 +587,18 @@ SmallVector<InterpreterValue> eval(Region &region,
...
+      auto startIndices = scope.findTensor(op.getStartIndices());
+      auto sliceSizes = scope.findTensor(op.getSliceSizes());
+      auto result = gatherOp(
+          operand, startIndices, Axes(op.getDimensionNumbers().getOffsetDims()),
+          Axes(op.getDimensionNumbers().getCollapsedSliceDims()),
+          Axes(op.getDimensionNumbers().getStartIndexMap()),
+          Axis(op.getDimensionNumbers().getIndexVectorDim()),
+          makeSizes(sliceSizes), op.getIndicesAreSorted(), op.getType());
+      scope.add(op.getResult(), result);
+
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply stablehlo/reference/Ops.cpp]: done

File: stablehlo/dialect/TypeInference.cpp [hunk 4]
```diff
@@ -2280,12 +2280,35 @@ LogicalResult inferDynamicGatherOp(
...
+        return emitOptionalError(location, "slice_sizes collapsed dimension ",
+                                 dim, " should <= 1 but got ", sliceDimSize);
+    }
+
+    // dynamic_gather_c12
+    for (const auto& it : llvm::enumerate(sliceSizesValues)) {
+      if (operandShape.isDynamicDim(it.index())) continue;
+      auto operandDimSize = operandShape.getDimSize(it.index());
+      auto sliceDimSize = it.value();
+      if (sliceDimSize < 0 || sliceDimSize > operandDimSize)
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply stablehlo/dialect/TypeInference.cpp]: done. Good catch!

File: stablehlo/dialect/TypeInference.cpp [hunk 5]
```diff
@@ -2280,12 +2280,35 @@ LogicalResult inferDynamicGatherOp(
...
+    // dynamic_gather_c8
+    for (auto dim : collapsedSliceDims) {
+      int64_t sliceDimSize = sliceSizesValues[dim];
+      if (sliceDimSize > 1)
+        return emitOptionalError(location, "slice_sizes collapsed dimension ",
+                                 dim, " should <= 1 but got ", sliceDimSize);
+    }
+
+    // dynamic_gather_c12
+    for (const auto& it : llvm::enumerate(sliceSizesValues)) {
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply stablehlo/dialect/TypeInference.cpp]: done.

File: stablehlo/dialect/TypeInference.cpp (hunk 1 above)
Comment (sdasgup3) [reply stablehlo/dialect/TypeInference.cpp]: good catch! done.

File: docs/spec.md [hunk 6]
```diff
@@ -2635,6 +2635,95 @@ planning to address this in
...
+// %operand: [
+//            [[1, 2], [3, 4], [5, 6], [7, 8]],
+//            [[9, 10],[11, 12], [13, 14], [15, 16]],
+//            [[17, 18], [19, 20], [21, 22], [23, 24]]
+//           ]
+// %start_indices: [
+//                  [[0, 0], [1, 0], [2, 1]],
+//                  [[0, 1], [1, 1], [0, 2]]
+//                 ]
+// %slize_sizes: [ 1, 2, 2]
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md]: done.

File: docs/spec.md [hunk 7]
```diff
@@ -2635,6 +2635,95 @@ planning to address this in
 
 &nbsp;[More Examples](https://github.com/openxla/stablehlo/tree/main/stablehlo/tests/interpret/dot_general.mlir)
 
+### dynamic_gather
+
+#### Semantics
+
+This operation is functionally identical to
+[iota](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#gather)
+op, with the `slice_sizes` specified dynamically as an operand.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md]: done.

File: stablehlo/dialect/StablehloOps.td [hunk 8]
```diff
@@ -3462,32 +3462,30 @@ def StableHLO_DynamicGatherOp: StableHLO_Op<"dynamic_gather",
...
 
   let arguments = (ins
-    HLO_Tensor:$operand,
-    HLO_IntTensor:$start_indices,
-    HLO_IntTensor:$slice_sizes,
-    StableHLO_GatherDimensionNumbers:$dimension_numbers,
-    DefaultValuedOptionalAttr<BoolAttr, "false">:$indices_are_sorted
+    HLO_Tensor:$operand /*dynamic_gather_i1*/,
+    HLO_IntTensor:$start_indices /*dynamic_gather_i2*/,
+    HLO_StaticDimensionTensor:$slice_sizes /*dynamic_gather_i3*/,
```
Comment (GleasonK) [inline stablehlo/dialect/StablehloOps.td:3486]: Why this change?

Comment (mlevesquedion) [reply stablehlo/dialect/StablehloOps.td:3486]: Otherwise the slice_sizes can be dynamic, but we don't want to allow that flexibility because that is only useful for an unranked operand (if the operand is ranked, then we should know the size of slice_sizes).

Comment (sdasgup3) [reply stablehlo/dialect/StablehloOps.td:3486]: This will allow the ODS to enforce that slize_size is 1-D for dynamic_gather op. Only this new here is the inclusion of `Index` type, but we are already using `Index` for things other that axis (e.g.  dynamic_pad op)

Comment (GleasonK) [reply stablehlo/dialect/StablehloOps.td:3486]: Let's not expand support for `index` I don't think? Leaves us in an inconsistent state.. I.e. why not for `start_indices`?

Comment (mlevesquedion) [reply stablehlo/dialect/StablehloOps.td:3486]: Ah good point. Looks like we need `HLO_StaticIntTensor` :upside_down_face:

Comment (sdasgup3) [reply stablehlo/dialect/StablehloOps.td:3486]: Agreed. `Static1DIntTensor` to be precise.

Comment (sdasgup3) [reply stablehlo/dialect/StablehloOps.td:3486]: Let us address this is a follow up PR.

File: stablehlo/dialect/StablehloOps.td [hunk 9]
```diff
@@ -3462,32 +3462,30 @@ def StableHLO_DynamicGatherOp: StableHLO_Op<"dynamic_gather",
...
+    %slice_sizes = stablehlo.constant dense<[1, 2, 2]> : tensor<3xi32>
     %result = "stablehlo.dynamic_gather"(%operand, %start_indices, %slice_sizes) {
       dimension_numbers = #stablehlo.gather<
         offset_dims = [2, 3],
         collapsed_slice_dims = [0],
         start_index_map = [0, 2],
         index_vector_dim = 2>,
       indices_are_sorted = false
-    } : (tensor<3x4x2xi32>, tensor<2x3x2xi64>, tensor<3xi64>) -> tensor<2x3x2x2xi32>
+    } : (tensor<3x4x2xi32>, tensor<2x3x2xi64>, tensor<3xi32>) -> tensor<2x3x2x2xi32>
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply stablehlo/dialect/StablehloOps.td:3479]: https://github.com/openxla/stablehlo/pull/2320

File: docs/spec.md [hunk 10]
```diff
@@ -2635,6 +2635,95 @@ planning to address this in
...
+//                 ]
+// %slize_sizes: [ 1, 2, 2]
+%result = "stablehlo.dynamic_gather"(%operand, %start_indices, %slize_sizes) {
+  dimension_numbers = #stablehlo.gather<
+    offset_dims = [2, 3],
+    collapsed_slice_dims = [0],
+    start_index_map = [1, 0],
+    index_vector_dim = 2>,
+  indices_are_sorted = false
+} : (tensor<3x4x2xi32>, tensor<2x3x2xi64>, tensor<3xi32>) -> tensor<2x3x2x2xi32>
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md:2710]: https://github.com/openxla/stablehlo/pull/2320

File: stablehlo/dialect/StablehloOps.td (hunk 8 above)
Comment (sdasgup3) [reply stablehlo/dialect/StablehloOps.td:3486]: https://github.com/openxla/stablehlo/pull/2320
